src/disk_io_thread_pool.cpp

```diff
@@ -0,0 +1,200 @@
+		, m_min_idle_threads(0)
+		, m_idle_timer(ios)
+	{}
+
+	disk_io_thread_pool::~disk_io_thread_pool()
+	{
+		abort(true);
+	}
+
+	void disk_io_thread_pool::set_max_threads(int const i)
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (i == m_max_threads) return;
+		m_max_threads = i;
+		if (m_threads.size() < i) return;
+		stop_threads(int(m_threads.size()) - i);
+	}
+
+	void disk_io_thread_pool::abort(bool wait)
+	{
+		std::unique_lock<std::mutex> l(m_mutex);
+		if (m_abort) return;
+		m_max_threads = 0;
+		m_abort = true;
+		m_idle_timer.cancel();
+		stop_threads(int(m_threads.size()));
+		for (auto& t : m_threads)
+		{
+			if (wait)
+			{
+				// must release m_mutex to avoid a deadlock if the thread
+				// tries to aquire it
+				l.unlock();
+				t.join();
+				l.lock();
+			}
+			else
+				t.detach();
+		}
+		m_threads.clear();
+	}
+
+	void disk_io_thread_pool::thread_active()
+	{
+		--m_num_idle_threads;
+		TORRENT_ASSERT(m_num_idle_threads >= 0);
+
+		int current_min = m_min_idle_threads;
+		while (m_num_idle_threads < current_min
+			&& !m_min_idle_threads.compare_exchange_weak(current_min, m_num_idle_threads));
+	}
+
+	bool disk_io_thread_pool::try_thread_exit(std::thread::id id)
+	{
+		int to_exit = m_threads_to_exit;
+		while (to_exit > 0 &&
+			!m_threads_to_exit.compare_exchange_weak(to_exit, to_exit - 1));
+		if (to_exit > 0)
+		{
+			std::unique_lock<std::mutex> l(m_mutex);
+			if (!m_abort)
+			{
+				auto new_end = std::remove_if(m_threads.begin(), m_threads.end()
+					, [id, &l](std::thread& t)
+				{
+					if (t.get_id() == id)
+					{
+						t.detach();
+						return true;
+					}
+					return false;
+				});
+				TORRENT_ASSERT(new_end != m_threads.end());
+				m_threads.erase(new_end, m_threads.end());
+				if (m_threads.empty()) m_idle_timer.cancel();
+			}
+		}
+		return to_exit > 0;
+	}
+
+	std::thread::id disk_io_thread_pool::first_thread_id()
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_threads.empty()) return std::thread::id();
+		return m_threads.front().get_id();
+	}
+
+	void disk_io_thread_pool::job_queued(int queue_size)
+	{
+		// this check is not strictly necessary
+		// but do it to avoid aquiring the mutex in the trivial case
+		if (m_num_idle_threads >= queue_size) return;
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_abort) return;
+
+		// reduce the number of threads requested to stop if we're going to need
+		// them for these new jobs
+		int to_exit = m_threads_to_exit;
+		while (to_exit > (std::max)(0, m_num_idle_threads - queue_size) &&
+			!m_threads_to_exit.compare_exchange_weak(to_exit
+				, (std::max)(0, m_num_idle_threads - queue_size)));
+
+		// now start threads until we either have enough to service
+		// all queued jobs without blocking or hit the max
+		for (int i = m_num_idle_threads
+			; i < queue_size && int(m_threads.size()) < m_max_threads
+			; ++i)
+		{
+			// if this is the first thread started, start the reaper timer
+			if (m_threads.empty())
```

Comment:
There is an inherent race if a reaper event is queued just before the timer is canceled. This isn't a fatal error, if the reaper runs back-to-back it will just cause a request for all idle threads at that moment to exit. So at worst it's a minor performance hit.